Add typing to style guide and update a few examples

sopelj opened the pull request.

fblackburn1 reviewed (commented):
Nice example
About the refactoring/styling part, what do you thing about removing good/bad practice already handled by our linters? Is it worth keeping them here?

sopelj commented:
> Nice example About the refactoring/styling part, what do you thing about removing good/bad practice already handled by our linters? Is it worth keeping them here?

Thanks! Yeah, good point. I'm up for removing things. Did you have some in mind? Black is very opinionated it covers a few of them. It also covers a bunch of things we don't mention. We could always point people to black or pep8 for more info for some points

fblackburn1 commented:
> > Nice example About the refactoring/styling part, what do you thing about removing good/bad practice already handled by our linters? Is it worth keeping them here?
> 
> Thanks! Yeah, good point. I'm up for removing things. Did you have some in mind? Black is very opinionated it covers a few of them. It also covers a bunch of things we don't mention. We could always point people to black or pep8 for more info for some points

:+1: for everything, if you want to point black/pep8 too
I don't have specific item in mind. I just know that these examples are here long time before we got linters

sopelj commented:
> > > Nice example About the refactoring/styling part, what do you thing about removing good/bad practice already handled by our linters? Is it worth keeping them here?
> > 
> > 
> > Thanks! Yeah, good point. I'm up for removing things. Did you have some in mind? Black is very opinionated it covers a few of them. It also covers a bunch of things we don't mention. We could always point people to black or pep8 for more info for some points
> 
> +1 for everything, if you want to point black/pep8 too I don't have specific item in mind. I just know that these examples are here long time before we got linters and CI inside PR

Ok, thanks! I added a section about linters with both tox and pre-commit and removed the examples with backslashes and redundant parentheses as they are covered by black.

fblackburn1 commented:
LGTM, I'll like a quick review of @DrPyser if there is something missing or to be improved

fblackburn1 commented:
LGTM

sopelj commented on content/uc-doc/contributors/typing.md:
```diff
@@ -0,0 +1,497 @@
+You can also use `if TYPE_CHECKING:` to encompass code that is only necessary for type checking or
+that you don’t want to run at runtime.
+
+This allows you to for example:
```
Missing a comma, but this is perfectly valid :P

sopelj commented on content/uc-doc/contributors/style_guide.md:
```diff
@@ -406,3 +370,61 @@ except UserNotFoundError as e:
+
+We use [`mypy`](https://mypy.readthedocs.io/en/stable/) to do type checking, and it is run via
+[`pre-commit`](https://pre-commit.com/). It can either be run automatically as a git pre-commit hook
+(after `pre-commit install` in your repo), manually with `pre-commit run --all-files` or via tox
```
I didn't put mypy after since there is no equivalent for tox. It is run via linters with all the others.

sopelj commented on content/uc-doc/contributors/style_guide.md:
````diff
@@ -406,3 +370,61 @@ except UserNotFoundError as e:
+```python
+from typing import TypedDict, Union
+
+UserData = TypedDict('UserData', {'first_name': str, 'last_name': str, 'email': Union[str, None]})
````
It's still quite readable, but it was mostly to avoid huge blocks in the if type checking. Just trying to be concise.

sopelj replied:
Also the `|` syntax doesn't work for assignment as is uses them to construct the dict.

sopelj commented on content/uc-doc/contributors/style_guide.md:
```diff
@@ -34,127 +34,83 @@ import re
+### General Style Rules
 
-When possible, use black to validate your code.
+To try and maintain a clean and consistent code base we use `black` which is a tool that enforces a
```
I think it's clear that black is a tool from context. Also, it is not only for style. Ity enforces consistent code. :)

sopelj commented on content/uc-doc/contributors/typing.md:
```diff
@@ -0,0 +1,497 @@
+title: Type Hinting
+---
+
+Here are some examples and basics on typing. More information on the why and how to run type
```
It would be "basic typing principles" if you wanted to say it that way, but I don't think it's required.

sopelj commented on content/uc-doc/contributors/typing.md:
```diff
@@ -0,0 +1,497 @@
+### Literal values {#literals}
+
+To allow only a limited selection of values you can use
+`[Literal](https://docs.python.org/3/library/typing.html#typing.Literal)`. It is useful for when you
+expect specific strings like `'on'` or `'off'`. The standard library uses this for read/write modes
```
They are quite different than Enum classes. They are simply distinct string values. Not sure of the link here

sopelj commented on content/uc-doc/contributors/typing.md:
```diff
@@ -0,0 +1,497 @@
+Mixin can be challenging to type if they have implicit dependency on other interfaces (i.e. expect
+the inheriting class to implement other methods used by the mixin implementation). Intersection
+types are not supported yet, but would allow specifying a self-type that combines the mixin’s
+interface.
```
No, it's singular

sopelj commented on content/uc-doc/contributors/typing.md:
```diff
@@ -0,0 +1,497 @@
+### Tox {#tox}
+
+We can also use `tox` to run mypy with our other linters via `pre-commit` this allows for a unified
+workflow, and it is run in our CI.
```
It is not just consistent, it is literally run.
